### DISCUSSION TITLE
SQLAlchemy with web frameworks.

### DISCUSSION THREAD

**Discussion by capymind** [sqlalchemy/sqlalchemy#10734] [Usage Questions] [answered]
Hi, first of all.
I am not good at Python.. sorry in advance this question doesn't make sense at all.

There are many web frameworks in Python ecosystem.
Flask, Quart, FastAPI, Starlette, Litestar, to name a few.

And they all have a kind of extension to glue SQLAlchemy with themselves.
Such as Flask-SQLAlchemy, Quart-SQLAlchemy, etc. 
Although there is no extension(library), 
there is guide line or how-to code to provide SQLAlchemy session, connection etc.

These are all great and help many users. However there are some issues that
I can not understand in each github discussion/issues/pull request etc. (my skill/knowledge issue)

Just to create (async/sync) engine and use it with context manager all the time is not enough to use
SQLAlchemy in web framework(request-response cycle)?

For instance

```python
engine = create_async_engine("...")

@webframework.route("/some-path")
def handle_request(...):
    with engine.begin() as conn:
        # some action

    with AsyncSession(engine) as session:
        # sodm action
```
In order to get knowledge about important thing related to SQLAlchmey with Webframeworks,
or why the code snippet above is not enough

I think the definitions of async, thread, session etc.. are needed to correctly understood..
Is there way to get(learn) these kinds of thing?

Sorry for this naive question..
SQAlchemy is really cool and gives me the opportunity to learn many database/orm related concepts.
Thanks all for precious hard works.

**Comment by CaselIT** [comment, marked as answer]
Hi,

The snippet you provided should in general work with sqlalchemy, since it will use a different connection/session in each request. This is kind the main point to take into consideration when using sqlalchemy with web frameworks

Note that if you use an async engine you will need to use `async` in the `def` and `with` construct (python requires them to distinguish between the sync and the async version)

In general what these extension do is provide a way of better integrating the the web framework in use with sqlalchemy, usually automatically providing to you a session that's committed or rolled back at the end of the request depending on the status. But they are just convenience tools that are optional.

>  Is there way to get(learn) these kinds of thing?

It depends on what you are looking for. If it's more information about sqlalchemy the tutorial is a great place to start https://docs.sqlalchemy.org/en/20/tutorial/index.html.
This may be of interest to understand how to manage the life of a session https://docs.sqlalchemy.org/en/20/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it

If it's more information regarding how to use sqlalchemy with a particular web framework, I don't have much sorry. I would suggest looking at the documentation of the framework you are using to see if they have something.

**Reply by capymind** [reply]
> Note that if you use an async engine you will need to use async in the def and with construct (python requires them to distinguish between the sync and the async version)

Yes.. you're totally right. Thank you for pointing me out. I'm mixing it in wrong way many times.. (8-9 of 10).

> But they are just convenience tools that are optional.

That's what I really want to hear from. 
Without solid(deep) knowledge of web framework and sqlalchemy(both are huge by definition!), it seems like for me that if I don't use extension or related library, I'm in problem very easily without noticing it..

> The snippet you provided should in general work with sqlalchemy, since it will use a different connection/session in each request. This is kind the main point to take into consideration when using sqlalchemy with web frameworks

I will take it into account. 
Thank you for shaing your knowledge and useful links, helping me a lot.
